File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_cpp/code_output/PageUtil.cpp
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>

class PageUtil {
public:
    struct PageInfo {
        int currentPage;
        int perPage;
        int totalPages;
        int totalItems;
        bool hasPrevious;
        bool hasNext;
        std::vector<int> data;

        PageInfo(int currentPage, int perPage, int totalPages, int totalItems,
                 bool hasPrevious, bool hasNext, const std::vector<int>& data)
            : currentPage(currentPage), perPage(perPage), totalPages(totalPages),
              totalItems(totalItems), hasPrevious(hasPrevious), hasNext(hasNext),
              data(data) {}

        bool operator==(const PageInfo& other) const {
            return currentPage == other.currentPage &&
                   perPage == other.perPage &&
                   totalPages == other.totalPages &&
                   totalItems == other.totalItems &&
                   hasPrevious == other.hasPrevious &&
                   hasNext == other.hasNext &&
                   data == other.data;
        }

        int hashCode() const {
            int result = currentPage;
            result = 31 * result + perPage;
            result = 31 * result + totalPages;
            result = 31 * result + totalItems;
            result = 31 * result + (hasPrevious ? 1 : 0);
            result = 31 * result + (hasNext ? 1 : 0);
            int dataHash = 1;
            for (int item : data) {
                dataHash = 31 * dataHash + item;
            }
            result = 31 * result + dataHash;
            return result;
        }
    };

    struct SearchResult {
        std::string keyword;
        int totalResults;
        int totalPages;
        std::vector<int> results;

        SearchResult(const std::string& keyword, int totalResults, int totalPages,
                     const std::vector<int>& results)
            : keyword(keyword), totalResults(totalResults), totalPages(totalPages),
              results(results) {}

        bool operator==(const SearchResult& other) const {
            return totalResults == other.totalResults &&
                   totalPages == other.totalPages &&
                   keyword == other.keyword &&
                   results == other.results;
        }

        int hashCode() const {
            int keywordHash = 0;
            for (char c : keyword) {
                keywordHash = 31 * keywordHash + static_cast<int>(c);
            }
            int result = keywordHash;
            result = 31 * result + totalResults;
            result = 31 * result + totalPages;
            int resultsHash = 1;
            for (int item : results) {
                resultsHash = 31 * resultsHash + item;
            }
            result = 31 * result + resultsHash;
            return result;
        }
    };

private:
    std::vector<int> data;
    int pageSize;
    int totalItems;
    int totalPages;

public:
    PageUtil(const std::vector<int>& data, int pageSize)
        : data(data), pageSize(pageSize),
          totalItems(static_cast<int>(data.size())),
          totalPages(static_cast<int>(std::ceil(static_cast<double>(totalItems) / pageSize))) {}

    std::vector<int> getPage(int pageNumber) {
        if (pageNumber < 1 || pageNumber > totalPages) {
            return std::vector<int>();
        }
        int startIndex = (pageNumber - 1) * pageSize;
        int endIndex = std::min(startIndex + pageSize, totalItems);
        return std::vector<int>(data.begin() + startIndex, data.begin() + endIndex);
    }

    PageInfo getPageInfo(int pageNumber) {
        if (pageNumber < 1 || pageNumber > totalPages) {
            return PageInfo(0, pageSize, totalPages, totalItems, false, false, std::vector<int>());
        }
        int startIndex = (pageNumber - 1) * pageSize;
        int endIndex = std::min(startIndex + pageSize, totalItems);
        std::vector<int> pageData(data.begin() + startIndex, data.begin() + endIndex);
        bool hasPrevious = pageNumber > 1;
        bool hasNext = pageNumber < totalPages;
        return PageInfo(pageNumber, pageSize, totalPages, totalItems, hasPrevious, hasNext, pageData);
    }

    SearchResult search(const std::string& keyword) {
        std::vector<int> results;
        for (int item : data) {
            if (std::to_string(item).find(keyword) != std::string::npos) {
                results.push_back(item);
            }
        }
        int numResults = static_cast<int>(results.size());
        int numPages = static_cast<int>(std::ceil(static_cast<double>(numResults) / pageSize));
        return SearchResult(keyword, numResults, numPages, results);
    }
};

File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_py/code_output/PageUtil.py
import math
from typing import List, Optional


class PageInfo:
    def __init__(self, current_page: int, per_page: int, total_pages: int,
                 total_items: int, has_previous: bool, has_next: bool,
                 data: Optional[List[int]] = None):
        self.current_page = current_page
        self.per_page = per_page
        self.total_pages = total_pages
        self.total_items = total_items
        self.has_previous = has_previous
        self.has_next = has_next
        self.data = list(data) if data is not None else []

    def __eq__(self, obj):
        if self is obj:
            return True
        if obj is None or type(self) is not type(obj):
            return False
        other = obj
        return (self.current_page == other.current_page and
                self.per_page == other.per_page and
                self.total_pages == other.total_pages and
                self.total_items == other.total_items and
                self.has_previous == other.has_previous and
                self.has_next == other.has_next and
                self.data == other.data)

    def __hash__(self):
        result = self.current_page
        result = 31 * result + self.per_page
        result = 31 * result + self.total_pages
        result = 31 * result + self.total_items
        result = 31 * result + (1 if self.has_previous else 0)
        result = 31 * result + (1 if self.has_next else 0)
        result = 31 * result + hash(tuple(self.data))
        return result


class SearchResult:
    def __init__(self, keyword: str, total_results: int, total_pages: int,
                 results: Optional[List[int]] = None):
        self.keyword = keyword
        self.total_results = total_results
        self.total_pages = total_pages
        self.results = list(results) if results is not None else []

    def __eq__(self, obj):
        if self is obj:
            return True
        if obj is None or type(self) is not type(obj):
            return False
        other = obj
        return (self.total_results == other.total_results and
                self.total_pages == other.total_pages and
                self.keyword == other.keyword and
                self.results == other.results)

    def __hash__(self):
        result = hash(self.keyword)
        result = 31 * result + self.total_results
        result = 31 * result + self.total_pages
        result = 31 * result + hash(tuple(self.results))
        return result


class PageUtil:
    def __init__(self, data: List[int], page_size: int):
        self.data = data
        self.page_size = page_size
        self.total_items = len(data)
        self.total_pages = math.ceil(self.total_items / page_size)

    def get_page(self, page_number: int) -> List[int]:
        if page_number < 1 or page_number > self.total_pages:
            return []
        start_index = (page_number - 1) * self.page_size
        end_index = min(start_index + self.page_size, self.total_items)
        return self.data[start_index:end_index]

    def get_page_info(self, page_number: int) -> PageInfo:
        if page_number < 1 or page_number > self.total_pages:
            return PageInfo(0, self.page_size, self.total_pages, self.total_items,
                            False, False, [])
        start_index = (page_number - 1) * self.page_size
        end_index = min(start_index + self.page_size, self.total_items)
        page_data = self.data[start_index:end_index]
        has_previous = page_number > 1
        has_next = page_number < self.total_pages
        return PageInfo(
            page_number,
            self.page_size,
            self.total_pages,
            self.total_items,
            has_previous,
            has_next,
            page_data
        )

    def search(self, keyword: str) -> SearchResult:
        results = [item for item in self.data if keyword in str(item)]
        num_results = len(results)
        num_pages = math.ceil(num_results / self.page_size)
        return SearchResult(
            keyword,
            num_results,
            num_pages,
            results
        )